Record Metrics for Reminder

Vyom-Yadav opened the pull request:
# Summary

***Provide a brief overview of the changes and the issue being addressed. 
Explain the rationale and any background necessary for understanding the changes. 
List dependencies required by this change, if any.***

Fixes #(related issue)

## Change Type

***Mark the type of change your PR introduces:***

- [ ] Bug fix (resolves an issue without affecting existing features)
- [x] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [ ] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

***Outline how the changes were tested, including steps to reproduce and any relevant configurations. 
Attach screenshots if helpful.***

# Review Checklist:

- [ ] Reviewed my own code for quality and clarity.
- [ ] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [ ] Included tests that validate the fix or feature.
- [ ] Checked that related changes are merged.

evankanderson commented on internal/reminder/metrics/provider.go:
```diff
@@ -0,0 +1,132 @@
+
+	// Create Prometheus exporter
+	prometheusExporter, err := prometheus.New(
+		prometheus.WithNamespace("reminder_service"),
```
```suggestion
		prometheus.WithNamespace("reminder"),
```

Skip "service" here -- it's implicit in the context, and the namespace may get stuck onto a bunch of metric names, and make them even longer and more difficult to read.  😁

evankanderson commented on pkg/config/reminder/metrics.go:
```diff
@@ -0,0 +1,7 @@
+
+type MetricsConfig struct {
+	Enabled bool   `mapstructure:"enabled" default:"true"`
+	Host    string `mapstructure:"host" default:"127.0.0.1"`
```
You may want to leave the address as `""`, so that Prometheus (for example) running in another network namespace can reach the metrics endpoint.

Vyom-Yadav replied:
I have configured this to localhost by default (changed the port). Can be overridden with configuration.

evankanderson commented on internal/reminder/metrics/provider.go:
```diff
@@ -0,0 +1,132 @@
+	// Create resource with service information
+	res := resource.NewWithAttributes(
+		semconv.SchemaURL,
+		semconv.ServiceName("reminder-service"),
```
Ditto on removing "-service" noise.
```suggestion
		semconv.ServiceName("reminder"),
```

evankanderson commented on internal/reminder/metrics/provider.go:
```diff
@@ -0,0 +1,132 @@
+	res := resource.NewWithAttributes(
+		semconv.SchemaURL,
+		semconv.ServiceName("reminder-service"),
+		semconv.ServiceVersion("v0.1.0"),
```
Can you copy over the:

```
// TODO: Make this auto-generated
```

if we're going to include it.  😁   (I think this was my fault.)

evankanderson commented on internal/reminder/metrics/provider.go:
```diff
@@ -0,0 +1,132 @@
+	// Set global meter provider
+	otel.SetMeterProvider(mp)
+
+	// Create metrics
+	meter := mp.Meter("reminder-service")
+	metrics, err := NewMetrics(meter)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create metrics: %w", err)
+	}
```
If you break the setup out into a static function, I'd probably have the calling reminder code look like:

```
if err := setupMetrics(ctx, cfg); err != nil {
  return err
}
reminder.metrics, err = NewMetrics(otel.Meter("reminder"))
if err != nil {
  return err
}
```

That separates the "system setup" of the metrics exporter from the provisioning of metrics for the main reminder loop, and makes it easier in the future to test the metrics without accidentally setting up a whole prometheus exporter.

evankanderson commented on internal/reminder/metrics/provider.go:
```diff
@@ -0,0 +1,132 @@
+	otel.SetMeterProvider(mp)
+
+	// Create metrics
+	meter := mp.Meter("reminder-service")
```
```suggestion
	meter := mp.Meter("reminder")
```

evankanderson commented on internal/reminder/metrics/provider.go:
```diff
@@ -0,0 +1,132 @@
+}
+
+// Shutdown gracefully shuts down the metrics server
+func (p *Provider) Shutdown(ctx context.Context) error {
```
If you make this method private, then cancelling the context to `Start` or `NewProvider` becomes the only way to reach this code, which simplifies your lifecycle over having multiple callers.

evankanderson replied:
(among other things, you can guarantee that `p` is not `nil`, because you don't call the `shutdown` method unless `p` is non-nil.)

evankanderson commented on internal/reminder/metrics/provider.go:
```diff
@@ -0,0 +1,132 @@
+}
+
+// Metrics returns the metrics instance
+func (p *Provider) Metrics() *Metrics {
+	return p.metrics
+}
```
My main suggestion is to remove the `Provider` object in favor of a static setup function with cancellation, but if you do keep it, I'd suggesting making `Metrics` exported, rather than needing to add an accessor function.

evankanderson commented on internal/reminder/reminder.go:
```diff
@@ -143,6 +163,11 @@ func (r *reminder) sendReminders(ctx context.Context) error {
 
+	remMetrics := r.metricsProvider.Metrics()
+	if remMetrics != nil {
+		remMetrics.RecordBatch(ctx, int64(len(repos)))
```
If you make batches a histogram, this can just be a `Record()` on that histogram.

evankanderson commented on internal/reminder/metrics/metrics.go:
```diff
@@ -0,0 +1,84 @@
+	m.AvgBatchSize.Record(ctx, avgSize)
+}
+
+func (m *Metrics) RecordSendDelay(ctx context.Context, delaySeconds float64) {
+	m.SendDelay.Record(ctx, delaySeconds)
+}
```
It looks like this function isn't doing anything -- you're already exposing `SendDelay`, and there's not much semantic benefit to `foo.RecordSendDelay(...)` over `foo.SendDelay.Record(...)`.  That also saves you from needing to write a comment.

If you transform `BatchSize` into a histogram as well, then you might be able to get rid of all the methods on `Metrics`, which would be nice.

evankanderson reviewed (commented):
A bunch of comments, mostly aiming to help simplify your life.  😁

Vyom-Yadav commented:
Opening this, I'm working on this.

Vyom-Yadav commented on internal/reminder/reminder.go:
```diff
@@ -151,14 +173,19 @@ func (r *reminder) sendReminders(ctx context.Context) error {
 	repoIds := make([]uuid.UUID, len(repos))
 	for _, repo := range repos {
 		repoIds = append(repoIds, repo.ID)
+		if r.metrics != nil {
+			// sendDelay = Now() - ReminderLastSent - MinElapsed
+			reminderLastSent := repo.ReminderLastSent
+			if reminderLastSent.Valid {
+				r.metrics.SendDelay.Record(ctx, (time.Since(reminderLastSent.Time) - r.cfg.RecurrenceConfig.MinElapsed).Seconds())
```
I'm still not sure about this metric. Let's say a repo has been frequently updated every hour (due to GH webhook events) and `r.cfg.RecurrenceConfig.MinElapsed` is 2 hours. In this case, let's say after sending a reminder 18 hours back, every hour the repo was updated by GH webhook and wasn't eligible for a reminder. Now when it becomes eligible for the reminder `SendDelay`  would be recorded as `18h` which seems to be very wrong in this case. (I think we discussed this previously too)

Vyom-Yadav commented on internal/reminder/reminder.go:
```diff
@@ -151,14 +173,19 @@ func (r *reminder) sendReminders(ctx context.Context) error {
+			if reminderLastSent.Valid {
+				r.metrics.SendDelay.Record(ctx, (time.Since(reminderLastSent.Time) - r.cfg.RecurrenceConfig.MinElapsed).Seconds())
+			} else {
+				// Recording metric for first time reminders (after how long was the initial reminder sent)
+				newSendDelay := time.Since(repoToLastUpdated[repo.ID]) - r.cfg.RecurrenceConfig.MinElapsed
+				r.metrics.NewSendDelay.Record(ctx, newSendDelay.Seconds())
```
Added a new metric for first time reminders. (I'm thinking this is _the_ definition of `sendDelay` and what we have above is wrong)

Vyom-Yadav reviewed (commented):
@evankanderson Because we reused eventer code in https://github.com/mindersec/minder/commit/a6d9b2295746fed45ad7fef0c24b021d3a7b1a6a, which created `pubWithMetrics` using `minder` namespace, there are some non-reminder metrics being outputted by reminder (e.g. `minder_eventer_publish_time_seconds_bucket`). 

I think this can cause problems, but I assume them to use the same infra, so not sure about how big of a problem that is. (Is this something we want to tackle in this PR?)

Vyom-Yadav commented on internal/reminder/metrics/metrics.go:
```diff
@@ -0,0 +1,84 @@
+	// For tracking average calculation
+	// TODO: consider persisting this to avoid reset on restart (maybe)
+	totalBatches   int64
+	totalReminders int64
```
Done.

Vyom-Yadav commented on internal/reminder/metrics/metrics.go:
```diff
@@ -0,0 +1,84 @@
+	SendDelay metric.Float64Histogram
+
+	// Current number of reminders in the batch
+	BatchSize metric.Int64Gauge
```
Done.

Vyom-Yadav commented on internal/reminder/metrics/metrics.go:
```diff
@@ -0,0 +1,84 @@
+	160,  // 2m 40s
+	320,  // 5m 20s
+	640,  // 10m 40s
+	1280, // 21m 20s
```
Updated with:

```go
var delayBuckets = []float64{
	60,    // 1 minute
	300,   // 5 minutes
	600,   // 10 minutes
	1800,  // 30 minutes
	3600,  // 1 hour
	7200,  // 2 hours
	10800, // 3 hours
	18000, // 5 hours
	25200, // 7 hours
	36000, // 10 hours
}
```

Vyom-Yadav commented on internal/reminder/metrics/metrics.go:
```diff
@@ -0,0 +1,84 @@
+
+	// Update running average
+	m.totalBatches++
+	m.totalReminders += size
```
Good point, I think this should be doable now for multiple reminder processes as we can aggregate metrics which are being exposed.

Vyom-Yadav commented on internal/reminder/reminder.go:
```diff
@@ -151,14 +176,19 @@ func (r *reminder) sendReminders(ctx context.Context) error {
+				remMetrics.SendDelay.Record(ctx, (time.Now().Sub(reminderLastSent.Time) - r.cfg.RecurrenceConfig.MinElapsed).Seconds())
+			} else {
+				// TODO: Should the send delay be zero if the reminder has never been sent?
+				remMetrics.SendDelay.Record(ctx, 0)
```
Exposing a new metric now.

evankanderson reviewed (commented):
A few notes -- I think you're right about determining the metrics for "sendDelay".

If you want to be particularly cute, you could write:

```golang
sendDelay := time.Since(...) - ...MinElapsed
recorder := r.metrics.SendDelay
if !reminderLastSent.Valid {
  // Use a different metric for first-time reminders
  recorder = r.metrics.NewSendDelay
}
recorder.Record(ctx, sendDelay.Seconds())
```

Vyom-Yadav commented on internal/reminder/reminder.go:
```diff
@@ -80,6 +84,20 @@ func (r *reminder) Start(ctx context.Context) error {
+		}()
+
+		var err error
+		r.metrics, err = metrics.NewMetrics(otel.Meter("reminder"))
```
No, you should feel strongly. There is a race condition here: (I wouldn't have checked until you pointed 🥇 )

```go
	if r.cfg.MetricsConfig.Enabled {
		go func() {
			if err := r.startMetricServer(ctx); err != nil {
				logger.Error().Err(err).Msg("error starting metrics server")
			}
		}()

		// Race condition here

		var err error
		r.metrics, err = metrics.NewMetrics(otel.Meter("reminder"))
		if err != nil {
			return err
		}
	}
```

`otel.SetMeterProvider(mp)` might not have been executed by the time `otel.Meter("reminder")` is run.

Vyom-Yadav replied:
```go
// If this is called before a global MeterProvider is registered the returned
// Meter will be a No-op implementation of a Meter. When a global MeterProvider
// is registered for the first time, the returned Meter, and all the
// instruments it has created or will create, are recreated automatically from
// the new MeterProvider.
```

Hmm, although technically, it is a race condition, but I think it should be fine (still I'll change it). Evan, can it be possible that few metrics get recorded on the no-op implementation, and then when the MeterProvider is registered for the first time, those metrics are lost?

evankanderson replied:
That's a reasonable implementation for OTel, in my opinion.

My main reaction was that you went through the process of creating a `MeterProvider` in `startMetricsServer` and then 2 lines later in the outer function you go through OTel globals to fish it out again.  If the code was more separated, I'd say the globals made sense, but at this proximity, I'd go for explicit over spooky action at a distance.

Vyom-Yadav commented on internal/reminder/metrics_server.go:
```diff
@@ -0,0 +1,84 @@
+		logger.Err(err).Msg("error shutting down metrics provider")
+	}
+
+	return server.Shutdown(shutdownCtx)
```
Yeah, the main goroutine goes home without waiting for his younger sibling, let me change it to send both of them back home together.

evankanderson replied:
As mentioned, I'm not too worried about hanging up in the middle of a metrics request, so don't contort the code too much to make it happen.  `exit()` is a _very_ effective garbage collector.

evankanderson replied:
Oh, now I see the other concern -- you may want to shut down the HTTP server _before_ the metrics provider.  Generally, I follow a LIFO order on teardown, with the idea that items constructed earlier probably don't depend on later items, but the other way may happen.

Vyom-Yadav commented on internal/reminder/metrics/metrics.go:
```diff
@@ -0,0 +1,105 @@
+		return nil, err
+	}
+
+	totalBatches, err := meter.Int64Counter(
+		"total_batches",
+		metric.WithDescription("Total number of batches processed"),
+	)
+	if err != nil {
+		return nil, err
+	}
```
`reminder_total_batches_total` and `reminder_batch_size_count` give the same result. `reminder_total_reminders_total` and `reminder_batch_size_sum` give the same result. So, I don't think those two extra metrics are useful. I've removed them.

Vyom-Yadav commented on internal/reminder/reminder.go:
```diff
@@ -213,10 +243,11 @@ func (r *reminder) getEligibleRepositories(ctx context.Context, repos []db.Repos
 	}
 
-	return eligibleRepos, nil
+	return eligibleRepos, eligibleReposLastUpdated, nil
```
I could just return `idToLastUpdate` as well, but that contained non eligible repos as well (so my initial thought was to create a new map of only eligible repos - helps in debugging with breakpoints - but it's fine either way). I've changed now to return `idToLastUpdate`.

evankanderson replied:
My thought was primarily to avoid a bunch of copies and allocations.  You could also clear out the entries from `idToLastUpdate` as you filter them if you don't want to return the extra data?

Vyom-Yadav commented on internal/reminder/reminder.go:
```diff
@@ -74,21 +78,52 @@ func (r *reminder) Start(ctx context.Context) error {
+		metricsProviderReady := make(chan struct{})
+
+		go func() {
+			if err := r.startMetricServer(ctx, metricsProviderReady); err != nil {
+				logger.Fatal().Err(err).Msg("failed to start metrics server")
+			}
```
This will call `exit()` even when `server.Shutdown(shutdownCtx)` return an error, is this correct? (This will prevent the deferred stop call)

evankanderson replied:
Since `startMetricServer` blocks, this seems reasonable.  The other option would be to have `startMetricServer` return a `chan error` and execute asynchronously within the function (i.e. move 68-86 into the goroutine started by `startMetricServer`).

Vyom-Yadav replied:
I feel the current option is fine.

I replaced `logger.Fatal().Err(err)` to be not `Fatal()`. OS will eventually clean up the socket and release the port.

Vyom-Yadav commented on internal/reminder/reminder.go:
```diff
@@ -74,21 +78,52 @@ func (r *reminder) Start(ctx context.Context) error {
+			close(metricsServerDone)
+		}()
+
+		select {
+		case <-metricsProviderReady:
+			var err error
+			r.metrics, err = metrics.NewMetrics(otel.Meter("reminder"))
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			logger.Info().Msg("reminder stopped")
+			return nil
+		}
```
I hope this isn't me too complicating stuff.

evankanderson replied:
I do think you're being overly conscientious for shutdown from `main`, but I'll allow it.

Vyom-Yadav replied:
Thank you, kindly!

Vyom-Yadav commented on internal/reminder/reminder.go:
```diff
@@ -74,21 +78,52 @@ func (r *reminder) Start(ctx context.Context) error {
 	for {
 		select {
 		case <-ctx.Done():
+			if r.cfg.MetricsConfig.Enabled {
+				<-metricsServerDone
+			}
```
Same for this one (making sure I'm not complicating things)

evankanderson replied:
Here, you might more easily do the following:

```golang
if r.cfg.MetricsConfig.Enabled {
  ...
} else {
  close(metricsServerDone)
}
```

And then you won't need the conditional here.

Vyom-Yadav replied:
I didn't understand that. The current shutdown sequence waits for the metrics server to be done. (waiting so that deferred call cannot be executed i.e. the `Stop()` call for reminder)

evankanderson replied:
If you `close(metricsServerDone)` in the `else` block on 111, 118-120 and 124-126 can simply do `<-metricsServerDone` whether or not `r.cfg.MetricsConfig.Enabled` is true or false, because either case will close the channel.

I'd also be okay with leaking the metrics HTTP server if it doesn't cause harm elsewhere.

Vyom-Yadav replied:
Yeah, good point. I implemented this change.

Vyom-Yadav commented on internal/reminder/reminder.go:
```diff
@@ -143,6 +178,10 @@ func (r *reminder) sendReminders(ctx context.Context) error {
 		return fmt.Errorf("error creating reminder messages: %w", err)
 	}
 
+	if r.metrics != nil {
+		r.metrics.RecordBatch(ctx, int64(len(repos)))
+	}
```
I could change this to a direct call, but have kept it as it is for now.

evankanderson replied:
I think the only advantage of using an indirect call is that you could hide the `if r.metrics != nil` inside the `RecordBatch` call if you use a pointer receiver.

Another option is to decide that metrics will simply always be on in Reminder -- I don't know if that will simplify the code, but I've always found the "maybe you don't want metrics" to be a strange notion in Minder that was there before I got here.

Vyom-Yadav replied:
I removed this function with a direct call. `if r.metrics != nil` is down in the code as well, so replacing it at one place breaks the consistency.

Vyom-Yadav commented on internal/reminder/reminder.go:
```diff
@@ -151,13 +190,16 @@ func (r *reminder) sendReminders(ctx context.Context) error {
 
-	// TODO: Collect Metrics
-	// Potential metrics:
-	// - Gauge: Number of reminders in the current batch
-	// - UpDownCounter: Average reminders sent per batch
-	// - Histogram: reminder_last_sent time distribution
+			recorder := r.metrics.SendDelay
+			if !repo.ReminderLastSent.Valid {
+				recorder = r.metrics.NewSendDelay
+			}
+			recorder.Record(ctx, sendDelay.Seconds())
+		}
+	}
```
Now this all brings back the question that how useful is `ReminderLastSent` metric that we store in the main DB? Is there any other application for it? Right now it's just being used as a boolean for checking whether a repo has been reconciled or not (through reminder)

evankanderson replied:
Ah, you mean the column in the database?  Given where we're at now, maybe we don't need it?

Vyom-Yadav replied:
Let's keep it for some time around, till we setup reminder with metrics to see if this can be potentially valuable. We can cleanup later if not. (Adding code requires more testing)

Vyom-Yadav commented:
@evankanderson Did you see this:

> Because we reused eventer code in https://github.com/mindersec/minder/commit/a6d9b2295746fed45ad7fef0c24b021d3a7b1a6a, which created pubWithMetrics using minder namespace, there are some non-reminder metrics being outputted by reminder (e.g. minder_eventer_publish_time_seconds_bucket).
> 
> I think this can cause problems, but I assume them to use the same infra, so not sure about how big of a problem that is. (Is this something we want to tackle in this PR?)

evankanderson commented:
> @evankanderson Did you see this:
> 
> > Because we reused eventer code in [a6d9b22](https://github.com/mindersec/minder/commit/a6d9b2295746fed45ad7fef0c24b021d3a7b1a6a), which created pubWithMetrics using minder namespace, there are some non-reminder metrics being outputted by reminder (e.g. minder_eventer_publish_time_seconds_bucket).
> > I think this can cause problems, but I assume them to use the same infra, so not sure about how big of a problem that is. (Is this something we want to tackle in this PR?)

This isn't actually a big problem, because the pod / job labels (and the OTel MetricProvider Resource, if you're using the Otel binary protocol) will be different.  In fact, using the same metric name across different components makes it easier to build a dashboard that lets you "zoom out" and "zoom in" on how different components are handling common concerns (like queue lengths, latencies, etc).

evankanderson commented on internal/reminder/metrics/metrics.go:
```diff
@@ -0,0 +1,79 @@
+	}, nil
+}
+
+// RecordBatch records the metrics for a batch of reminders
+func (m *Metrics) RecordBatch(ctx context.Context, size int64) {
+	m.BatchSize.Record(ctx, size)
+}
```
This is a super-minor nit, but it feels like you could remove this method, and just have callers invoke `m.BatchSize.Record(ctx, size)`.

Vyom-Yadav replied:
Yeah, I kept it initially, but I've removed it now.

Vyom-Yadav commented:
CI, flaking TestNatsChannel

```
❌ TestNatsChannel (5.09s)
      natschannel_test.go:95: Got {"msg":"hello"} from out1
      natschannel_test.go:95: Got {"msg":"hola"} from out1
      natschannel_test.go:100: timeout waiting for message 2
      natschannel_test.go:114: expected 4 messages, got 2: [0xc0003c1020 0xc000160240]
```

Vyom-Yadav commented:
Hey @evankanderson, sorry for the delay on this, I got caught with one thing after the other. I addressed the review comments, ptal and let me know if anything else needs to be changed. Thanks!

evankanderson commented on internal/reminder/reminder.go:
```diff
@@ -74,21 +78,52 @@ func (r *reminder) Start(ctx context.Context) error {
 		return errors.New("reminder stopped, cannot start again")
 	default:
 	}
+	defer r.Stop()
```
This is kind of odd, since `Start` doesn't return, and would normally be exited by calling `Stop()`.  Add a comment here on why this is needed?

(I'm not saying it's wrong, I'm saying to drop a hint to prevent future authors from deleting the line.

Vyom-Yadav replied:
This defer here is mostly to handle context cancellation part. Depending on the invoker to clean up reminder after context cancellation isn't something I wanted to do. I added a comment on top.

evankanderson commented on internal/reminder/reminder.go:
```diff
@@ -74,21 +78,52 @@ func (r *reminder) Start(ctx context.Context) error {
+	if r.cfg.MetricsConfig.Enabled {
+		metricsProviderReady := make(chan struct{})
+
+		go func() {
+			if err := r.startMetricServer(ctx, metricsProviderReady); err != nil {
+				logger.Err(err).Msg("failed to start metrics server")
+			}
+			close(metricsServerDone)
+		}()
```
It feels weird to me to end up needing to call `startMetricServer` in another thread and then deal with that, given that `startMetricServer` _also_ sets up a background thread (and then waits for it).

We end up with:

```
main
 \--> startMetricServer
          \--> http.ListenAndServe
```

When it feels like we should have:

```
main
\--> http.ListenAndServe
```

Possibly with a "closer" callback.

I only mention this because `Join`-ing goroutines is a bunch of code that distracts from the rest of what's going on, so I tend to prefer to minimize how often I need to block on a goroutine exiting.

Vyom-Yadav replied:
Yeah, that's a good point. It was confusing especially with the part of getting the otel meter provider ready and set. I modified the implementation to be like:

```
main
\--> http.ListenAndServer
\--> watch for stopping the metrics server
```

It can't be done in a single go routine, but I think this solution is better than what was there previously.

evankanderson commented on internal/reminder/metrics_server.go:
```diff
@@ -0,0 +1,87 @@
+	// Indicates that a global MeterProvider is available
+	close(mpReady)
+
+	mux := http.NewServeMux()
+	mux.Handle(metricsPath, promhttp.Handler())
```
Why do you need an `http.ServeMux` here, rather than simply passing `promhttp.Handler` to the `http.Server` below?

Vyom-Yadav replied:
Do we want to serve the metrics on the root path? I don't see any problem with that, but it's just that I've seen `/metrics` being commonly used.

evankanderson replied:
(On the other hand, I don't have a strong complaint about putting this at `/metrics`)

evankanderson reviewed (dismissed):
I'm willing to merge this as-is if you want, or you can try to simplify the startup code.  I can understand just wanting to be done with it, though.  :grin:

Vyom-Yadav commented:
>  I can understand just wanting to be done with it, though. 😁

Not really, your reviews always teach me something, so deeply appreciated 🙇🏼‍♂️

evankanderson commented:
I'm going to merge this, since I think it's as close to perfect as we're going to manage.
